azure-kusto-data/azure/kusto/data/__init__.py

```diff
@@ -1,4 +1,148 @@
+    class _Keywords(Enum):
+        data_source = "Data Source"
+        aad_user_id = "AAD User ID"
+        password = "Password"
+        application_client_id = "Application Client Id"
+        application_key = "Application Key"
+        authority_id = "Authority Id"
+
+    def __init__(self, connection_string, *args, **kwargs):
+        self._constructing = True
+        if connection_string is not None and '=' not in connection_string.split(';')[0]:
+            connection_string = 'Data Source=' + connection_string
+        for kvp_string in connection_string.split(';'):
+            kvp = kvp_string.split('=')
+            self[kvp[0]] = kvp[1]
+        super(KustoConnectionStringBuilder, self).__init__(*args, **kwargs)
+        self._is_valid()
+        self._constructing = False
+
+    def __setitem__(self, key, value):
+        try:
+            keyword = KustoConnectionStringBuilder._Keywords(key)
+        except ValueError:
+            raise ValueError("%s is not supported as an item in KustoConnectionStringBuilder" % key)
+        super(KustoConnectionStringBuilder, self).__setitem__(keyword, value)
+        if not self._constructing:
+            self._is_valid()
+
+    def update(self, *args, **kwargs):
+        if args:
+            if len(args) > 1:
+                raise TypeError("update expected at most 1 arguments, "
+                                "got %d" % len(args))
+            other = dict(args[0])
+            for key in other:
+                self[key] = other[key]
+        if kwargs:
+            for key in kwargs:
+                self[key] = kwargs[key]
+        self._is_valid()
+
+    def setdefault(self, key, value=None):
+        if key not in self:
+            self[key] = value
+        self._is_valid()
+        return self[key]
+
+    @classmethod
+    def with_aad_user_password_authentication(cls, connection_string, user_id, password):
+        _ensure_value_is_valid(user_id)
+        _ensure_value_is_valid(password)
+        return cls(connection_string,\
+            {KustoConnectionStringBuilder._Keywords.aad_user_id: user_id,\
+            KustoConnectionStringBuilder._Keywords.password: password})
+
+    @classmethod
+    def with_aad_application_key_authentication(cls, connection_string, application_client_id, application_key):
+        _ensure_value_is_valid(application_client_id)
+        _ensure_value_is_valid(application_key)
+        return cls(connection_string,\
+            {KustoConnectionStringBuilder._Keywords.application_client_id: application_client_id,\
+            KustoConnectionStringBuilder._Keywords.application_key: application_key})
+
+    @classmethod
+    def with_aad_device_authentication(cls, connection_string):
+        return cls(connection_string)
+    
+    @property
+    def data_source(self):
+        if KustoConnectionStringBuilder._Keywords.data_source in self:
+            return self[KustoConnectionStringBuilder._Keywords.data_source]
+        return None
```

Comment:
All functions and methods implicitly have a `return None` at the end of them, so this line is unnecessary.